beacon-chain/powchain/service.go

```diff
@@ -143,72 +143,51 @@ func (w *Web3Service) Status() error {
 	return nil
 }
 
-// initDataFromVRC calls the vrc contract and finds the deposit count
-// and deposit root.
-func (w *Web3Service) initDataFromVRC() error {
-	root, err := w.vrcCaller.GetDepositRoot(&bind.CallOpts{})
-	if err != nil {
-		return fmt.Errorf("could not retrieve deposit root %v", err)
-	}
-
-	w.depositRoot = root[:]
-	w.depositTrie = trie.NewDepositTrie()
+// LatestBlockNumber in the ETH1.0 chain.
+func (w *Web3Service) LatestBlockNumber() *big.Int {
+	return w.blockNumber
+}
 
-	return nil
+// LatestBlockHash in the ETH1.0 chain.
+func (w *Web3Service) LatestBlockHash() common.Hash {
+	return w.blockHash
 }
 
-// run subscribes to all the services for the ETH1.0 chain.
-func (w *Web3Service) run(done <-chan struct{}) {
-	if err := w.initDataFromVRC(); err != nil {
-		log.Errorf("Unable to retrieve data from VRC %v", err)
-		return
-	}
+// Client for interacting with the ETH1.0 chain.
+func (w *Web3Service) Client() Client {
+	return w.client
+}
 
-	headSub, err := w.reader.SubscribeNewHead(w.ctx, w.headerChan)
-	if err != nil {
-		log.Errorf("Unable to subscribe to incoming ETH1.0 chain headers: %v", err)
-		return
-	}
+// HasChainStartLogOccurred queries all logs in the deposit contract to verify
+// if ChainStart has occurred. If so, it returns true alongside the ChainStart timestamp.
+func (w *Web3Service) HasChainStartLogOccurred() (bool, time.Time, error) {
```

Comment:
Every time this function is called, we request for all the logs from the ETH 1.0 chain for the deposit Contract and then loop through it to find if chainstart occured. Once `n` , where n is the number of full deposits in the deposit contract grows to a large amount this will become a bottleneck. As a call to an ETH 1.0 node is an expensive operation, can we cache the results of this ?

When starting up the node , we have to call the ETH 1.0 node for all the logs of the deposit contract, but the second time we do it, we do not have to request for all the logs again. So lets say the last time we requested for logs we received all logs till block X( ETH 1.0 chain), the next time we do it, we can request logs from block X+1 . You can use the filter query for this

Reply:
Can we do this in a separate PR? PR is already big as is

Reply:
Yeah , that would be better. The diff is pretty big as it is